pmlc/dialect/pxa/transforms/stencil.cc

```diff
@@ -0,0 +1,160 @@
+      return;
+    }
+
+    // Get the in tensors for the reduce op.
+    Value reduceIn = reduceOp.val();
+    MulOperationType mulOpType = MulOperationType::NoneMulOpType;
+
+    // Make sure the in for the reduce is a result of a multiplication.
+    auto valDef = reduceIn.getDefiningOp();
+
+    if (valDef == nullptr) {
+      op.emitRemark("the source of the reduce operation is not defined in this block");
+      IVLOG(3, "the source of the reduce operation is not defined in this block");
+      return;
+    }
+
+    mlir::MulFOp mulfOp  = llvm::dyn_cast_or_null<mlir::MulFOp>(valDef);
+    mlir::MulIOp muliOp  = llvm::dyn_cast_or_null<mlir::MulIOp>(valDef);
+    if (mulfOp == nullptr && muliOp == nullptr) {
+      op.emitRemark("the source of the reduce is not a multiplication operation");
+      IVLOG(3, "The source of the reduce is not a multiplication operation");
+      return;
+    }
+
+    mlir::AffineLoadOp lhs;
+    mlir::AffineLoadOp rhs;
+    if (mulfOp != nullptr) {
+      mulOpType = MulOperationType::FloatTy;
+      lhs = llvm::dyn_cast_or_null<mlir::AffineLoadOp>(mulfOp.lhs().getDefiningOp());
+      rhs = llvm::dyn_cast_or_null<mlir::AffineLoadOp>(mulfOp.rhs().getDefiningOp());
+    }
+    else if (muliOp != nullptr) {
+      mulOpType = MulOperationType::IntTy;
+      lhs = llvm::dyn_cast_or_null<mlir::AffineLoadOp>(muliOp.lhs().getDefiningOp());
+      rhs = llvm::dyn_cast_or_null<mlir::AffineLoadOp>(muliOp.rhs().getDefiningOp());
+    }
+    else {
+      op.emitRemark("unhandled multiplication type in stenciling");
+      IVLOG(3, "unhandled multiplication type in stenciling");
+    }
+
+    // Now verify the types of the operands of the mulOp must be affine.load operations.
+    if (lhs == nullptr || rhs == nullptr) {
+      op.emitRemark("the lhs or rhs of the mul operation are not an affne.load operations");
+      IVLOG(3, "the lhs or rhs of the mul operation are not an affne.load operations");
+      return;
+    }
+
+    // TODO: Need a bit better liveness analysis here to make sure the parameters 
+    // of any of the above 4 operations are not used in operations with
+    // side effects - store, calls, etc.
+
+    // Fill the values for the in/out/type of multiplication, etc.
+    GemmOperationMatch gemmMatch = GemmOperationMatch();
+    gemmMatch.mulOpType = mulOpType;
+    gemmMatch.in1Op = lhs;
+    gemmMatch.in2Op = rhs;
+    gemmMatch.outOp = reduceOp;
+    ret = gemmMatch;
+    op.setAttr("is_gemm", builder.getUnitAttr());
+  });
+
+  return ret;
+}
+
+void Stencil::DoStenciling(mlir::AffineParallelOp op, mlir::MLIRContext* mlirContext) {
+  // Initialization
+  auto matchOpt = getGemmOperation(op, mlirContext);
+  if (!matchOpt.hasValue()) {
+    IVLOG(3, "Not a Gemm match.");
+    return;
+  }
+}
+
+void StencilPass::runOnFunction() {
+  auto func = getFunction();
+  auto context = &getContext();
+  func.walk([&](mlir::AffineParallelOp op) {
+    Stencil as;
+    as.DoStenciling(op, context);
```

Comment:
No need to pass context, since almost everything in MLIR can access is (ie op.getContext() works here).

Reply:
Passed to the constructor of Stencil now.